### PR TITLE
Fixes : https://github.com/huggingface/nanotron/issues/114

### PR DESCRIPTION
**Pull request by MekkCyber**
Implementation of 1.58bit LLM with Llama following the paper & handbook released by Microsoft :

https://github.com/microsoft/unilm/blob/master/bitnet/The-Era-of-1-bit-LLMs__Training_Tips_Code_FAQ.pdf

Here are The training results on 25B tokens : 

loss_curve

cc @NouamaneTazi @xrsrke @thomwolf

### CODE REVIEW

**Comment by MekkCyber** [comment]
I think it has to do with the batch size. During our latest experiment, we trained the 1.58 model on 100B tokens, and we managed to get a 2.8 loss after 25B tokens with a batch size of 1024

File: src/nanotron/parallel/tensor_parallel/nn.py [hunk 1]
```diff
@@ -297,3 +298,227 @@ def forward(self, input_ids: torch.Tensor) -> torch.Tensor:
...
+            pg=self.pg,
+            split_config=split_config,
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if not self.training : 
+            w = self.weight  # a weight tensor with shape [d, k]
+            w = w.to(torch.bfloat16)
+            w_scale = getattr(self, "weight_scale").data
+            x_norm = normalize(x, self.in_features)
```
Comment (gau-nernst) [inline src/nanotron/parallel/tensor_parallel/nn.py:404]: Shouldn't RMSNorm here have learnable weights and biases?

File: src/nanotron/parallel/tensor_parallel/nn.py [hunk 2]
```diff
@@ -297,3 +298,227 @@ def forward(self, input_ids: torch.Tensor) -> torch.Tensor:
...
+                input=x_quant,
+                weight=w,
+                bias=self.bias,
+                group=self.pg,
+                tp_mode=self.mode,
+                async_communication=self.async_communication,
+            ) / w_scale / x_scale
+        else : 
+            w = self.weight
+            x_norm = normalize(x, self.in_features)
```
Comment (gau-nernst) [inline src/nanotron/parallel/tensor_parallel/nn.py:416]: Shouldn't RMSNorm here have learnable weights and biases?

**Comment by hjc3613** [comment]
Hi, I fetched this pr, and fintune the llama 70B using tp=8 or pp=8, before training, I have convert the llama 70B into nanotron format using this method: https://github.com/huggingface/nanotron/pull/174, but got this error:
image
so, how to finetune the 70B model into 1.58bit from scratch? could you give me some suggestion please? thank you!

File: src/nanotron/parallel/tensor_parallel/nn.py (hunk 2 above)
Comment (gau-nernst) [reply src/nanotron/parallel/tensor_parallel/nn.py:416]: @MekkCyber May I know why this is marked as resolved? From my understanding of the training tips handbook, the new RMSNorm should have learnable weights as the usual RMSNorm layers. Is there a reason you left it out here? (as well as in the PR merged in huggingface/transformers).

**Comment by MekkCyber** [comment]
Hey @hjc3613 thanks for the report ! you don't have to be consitent with the convert config, it should work I will investigate that! Can you tell me from your side what is the content of models/qwen2.5-72b-instruct-nanotron/model/model/decoder/0/pp_block/MLPBitNet

**Comment by hjc3613** [comment]
> Hey @hjc3613 thanks for the report ! you don't have to be consitent with the convert config, it should work I will investigate that! Can you tell me from your side what is the content of models/qwen2.5-72b-instruct-nanotron/model/model/decoder/0/pp_block/MLPBitNet

thank you very much! 
My requirement is to successfully fine-tune a 70B model to a 1.58bit model, using either Qwen2.5 70B or LLama 70B - both have almost identical model architectures. According to the training code, I need to provide a dataset and a base model in nanotron format, which will then be fine-tuned to 1.58bit for inference. However, I only have a HF version of the base model, so following PR174, I'm first converting it to nanotron format. Before conversion, I need to set dp=1, pp=1, tp=1. After conversion, I'm using the code from PR180 for training. This is what I'm currently doing - if there are any unreasonable or unnecessary steps, please advise. Thank you sincerely. To my understanding, a 70B model must have pp>1 or tp>1, as a single GPU card cannot hold the complete model.

**Comment by Andrefty** [comment]
Hi @MekkCyber !
Similar to what @hjc3613 wanted to try, I am trying to fine-tune the [Llama3-8B-1.58-100B-tokens](https://huggingface.co/HF1BitLLM/Llama3-8B-1.58-100B-tokens) model on a downstream task. (I have been detailing my progress in [this](https://huggingface.co/HF1BitLLM/Llama3-8B-1.58-100B-tokens/discussions/7) Hugging Face discussion thread)

So far I think I managed to correctly unpack the weights of the model using the functions provided by you, but the [`convert_hf_to_nanotron` script from this PR](https://github.com/huggingface/nanotron/pull/174) seems to not copy the `weight_scale` layers into the nanotron-format model (even after messing around with the `quantization_config`), resulting in a model that is outputting garbage when using the `run_generate` script with it (as a sanity check).

I am unsure if this is expected and if I should continue with fine-tuning, having the model in this state. I would appreciate any help or guidance you could provide on this matter.

File: src/nanotron/parallel/tensor_parallel/nn.py (hunk 2 above)
Comment (BradZhone) [reply src/nanotron/parallel/tensor_parallel/nn.py:416]: > @MekkCyber May I know why this is marked as resolved? From my understanding of the training tips handbook, the new RMSNorm should have learnable weights as the usual RMSNorm layers. Is there a reason you left it out here? (as well as in the PR merged in huggingface/transformers).

Only normalize function without learnable weights can already work well, maybe you can refer to this https://arxiv.org/abs/2505.08823

**Comment by lonleyodd** [comment]
@MekkCyber 
image
Hi! I noticed in your normalization function, you're computing mean and std along dimensions (-2, -1). I'm wondering if this is intentional? Typically, LayerNorm only normalizes along the last dimension (-1). Could you clarify the reasoning behind normalizing across both hidden/batch_size and feature dimensions? Thanks!

**Comment by MekkCyber** [comment]
the function is not used iirc, i used it for experiments and it didn't give better results
